src/pveclib/vec_int64_ppc.h

```diff
@@ -2175,6 +2180,97 @@ vec_cmpud_any_ne (vui64_t a, vui64_t b)
   return (result);
 }
 
+/** \brief Vector Load with gather unsigned doubleword from offsets.
+ *
+ *  For each doubleword element [i] of vra, load the doubleword
+ *  element at *array+vra[i]. Merge those doubleword elements and
+ *  return the resulting vector.
+ *
+ *  \note As effective address calculation is modulo 64-bits, signed or
+ *  unsigned doubleword offsets are equivalent.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   |   12  | 1/cycle  |
+ *  |power9   |   11  | 1/cycle  |
+ *
+ *  @param array Pointer to array of unsigned doublewords.
+ *  @param vra Vector of doubleword offsets.
+ *  @return vector doubleword containing elements loaded from
+ *  *array+vra[0] and *array+vra[1].
+ */
+static inline
+vui64_t
+vec_lvgudo (unsigned long long int *array, vui64_t vra)
+{
+  vui64_t rese0, rese1;
+
+  rese0 = vec_vlxsdx (vra[VEC_DW_H], array);
+  rese1 = vec_vlxsdx (vra[VEC_DW_L], array);
+  return  vec_permdi (rese0, rese1, 0);
```

Comment:
This seems to work for P9 and newer GCC. But can get ugly for older GCC and older POWER targets. And I have not even looked at what Clang does yet. 

For example for POWER8 (without mtvsrdd) requires two mfvsrrd's and xxpermdi. So better with VSX loads and xxpermdi. Also GCC 7 (AT11) for POWER7 generates multiple store reloads (no mfvsrd or mtvsrd). The POWER9 code generation evolves through GCC 7-10. Likely need ASM for older GCC and -mcpu=power9 anyway.

This will take some time to sort out. But for now ASM offers more consistent results given the combinatorics of compilers, compiler releases, and PowerISA releases we are dealing with.

For now I have a pattern that works. We should save the turd polishing for later.

Reply:
Also this is very good example of operations that the compiler should handle directly (as builtin).

But trying out patterns like this is a prerequisite.